be/src/exprs/ngram.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// This file is based on code available under the Apache license here:
+//  https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsStringSimilarity.cpp
+
+#include "column/column_hash.h"
+#include "exprs/string_functions.h"
+namespace starrocks {
+static constexpr size_t max_string_size = 1 << 15;
+
+template <size_t N, bool case_insensitive, bool use_utf_8, class Gram>
+class NgramFunctionImpl {
+public:
+    // we restrict needle's size smaller than 2^16, so even if every gram in needle is the same as each other
+    // we still only need one uint16 to store its frequency
+    using NgramHash = uint16;
+    // uint16[2^16] can almost fit into L2
+    static constexpr size_t map_size = 1 << 16;
+
+    static NgramHash getAsciiHash(const Gram* ch) { return crc_hash_32(ch, N, CRC_HASH_SEED1) & (0xffffu); }
+
+    // for every gram of needle, we calculate its' hash value and store its' frequency in map, and return the number of gram in needle
```

Comment:
why do we need to store the frequency?

Reply:
because this hashmap is not accurate, may exist hash collision, so store the frequency instead of value for speed